File: Java/438. Find All Anagrams in a String.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<>();
        
        if(p.length() > s.length()){
            return ans;
        }
        
        int[] chars1 = new int[26];
        for(int i=0; i<p.length(); i++){
            chars1[p.charAt(i) - 'a']++;
        }
        
        int diff = p.length();
        int[] chars2 = new int[26];
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            chars2[c-'a']++;
            if(i >= diff){
                chars2[s.charAt(i-diff) - 'a']--;
            }
            
            if(Arrays.equals(chars1, chars2)){
                ans.add(i - diff + 1);
            }
        }
        return ans;
    }
    
}

File: Python3/438. Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s):
            return []
        
        chars1 = collections.Counter(p)
                    
        ans = []
        left = right = 0
        chars2 = collections.defaultdict(int)
        
        while right < len(s):
            c = s[right]
            right += 1
            
            # refresh the dict
            if c not in chars1:
                chars2.clear()
                left = right
            
            else:
                chars2[c] += 1
                # check if need to update answer first
                if chars1 == chars2:
                    ans.append(left)
                while chars2[c] > chars1[c]:
                    chars2[s[left]] -= 1
                    left += 1
                    if chars1 == chars2:
                        ans.append(left)
                #print(chars2) 
        
        return ans
                
            